### PR TITLE
os/bluestore/BlueFS: use 64K alloc_size on the shared device

### PR DESCRIPTION
**Pull request by liewegas**
The shared devices is susceptible to fragmentation, making the 1 MB bluefs allocations fail.  Use a smaller allocation size (64K) for the shared device, while keeping the same large allocations for DB and WAL.

### CODE REVIEW

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -2527,41 +2542,38 @@ int BlueFS::_allocate(uint8_t id, uint64_t len,
...
     if (!node->extents.empty() && node->extents.back().bdev == id) {
       hint = node->extents.back().end();
     }   
     extents.reserve(4);  // 4 should be (more than) enough for most allocations
-    alloc_len = alloc[id]->allocate(left, min_alloc_size, hint, &extents);
+    alloc_len = alloc[id]->allocate(round_up_to(len, alloc_size[id]),
+				    alloc_size[id], hint, &extents);
   }
-  if (alloc_len < 0 || alloc_len < (int64_t)left) {
+  if (alloc_len < 0 || alloc_len < (int64_t)len) {
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: nit: more correct is to check against round_up_to(len, alloc_size[id])) not len

**Comment by neha-ojha** [comment]
needs a rebase since https://github.com/ceph/ceph/pull/29425 merged

**Comment by ifed01** [comment]
needs rebase

File: src/common/options.cc [hunk 2]
```diff
@@ -4252,7 +4252,11 @@ std::vector<Option> get_global_options() {
 
     Option("bluefs_alloc_size", Option::TYPE_SIZE, Option::LEVEL_ADVANCED)
     .set_default(1_M)
-    .set_description(""),
+    .set_description("Allocation unit size for DB and WAL devices"),
+
+    Option("bluefs_shared_alloc_size", Option::TYPE_SIZE, Option::LEVEL_ADVANCED)
```
Comment (jdurgin) [inline src/common/options.cc]: add to bluestore docs/

**Comment by liewegas** [comment]
I'm heading out for the day, can you queue up a test?  Thanks!

**Comment by neha-ojha** [comment]
> I'm heading out for the day, can you queue up a test? Thanks!

sure!

**Comment by neha-ojha** [comment]
the make check failure in unittest_bluefs looks related

**Comment by neha-ojha** [comment]
http://pulpito.ceph.com/nojha-2019-08-08_00:28:04-rados-wip-bluefs-shared-alloc-distro-basic-smithi/ - lot of test failures due to `Floating point exception`

**Comment by liewegas** [comment]
fixed

**Comment by liewegas** [comment]
@neha-ojha should i cherry-pick 8128240b4c9a0c49c6664c509c08871e3ec292c1 here too?

**Comment by neha-ojha** [comment]
> @neha-ojha should i cherry-pick [8128240](https://github.com/ceph/ceph/commit/8128240b4c9a0c49c6664c509c08871e3ec292c1) here too?

don't think that'll be required, "bluefs_alloc_size" will still be 1Mb

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2019-08-08_08:04:33-rados-wip-29537-kefu-distro-basic-smithi/

**Comment by markhpc** [comment]
cbt errors are due to a regression introduced in https://github.com/ceph/cbt/pull/178.  We merged https://github.com/ceph/cbt/pull/182 yesterday which turned out to be an incomplete fix. https://github.com/ceph/cbt/pull/184 should be a better fix until we introduce a better settings hierarchy.

File: src/os/bluestore/BlueFS.cc [hunk 3]
```diff
@@ -1190,71 +1192,89 @@ int BlueFS::device_migrate_to_existing(
...
-	    old_ext.length,
-	    (char*)&buf.at(0),
-	    buffered);
-	  if (r != 0) {
-	    derr << __func__ << " failed to read 0x" << std::hex
-	      << old_ext.offset << "~" <<old_ext.length << std::dec
-	      << " from " << (int)dev_target << dendl;
-	    return -EIO;
-	  }
+	rewrite = true;
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: can we break the loop here?

**Comment by neha-ojha** [comment]
latest rados run: http://pulpito.ceph.com/nojha-2019-08-13_23:51:37-rados-wip-bluefs-shared-alloc-2018-08-13-distro-basic-smithi/

**Comment by smithfarm** [comment]
@liewegas @jdurgin @neha-ojha @vumrou I added a backports summary to the description of this PR. The current mimic backport #30219 differs from the nautilus backport #29739 . . . is that OK?

**Comment by smithfarm** [comment]
@vumrao ^

**Comment by smithfarm** [comment]
I see there are also trackers. . . uff. Will try to get this cleaned up.

**Comment by vumrao** [comment]
@smithfarm Yes, mimic/luminous have little different handling as Josh mentioned in his commit(b5de477d04cd5b978332bee9d49e8663eb6b766f). Looks like now all taken care. Thank you.

**Pull request merged**